error.ts

```diff
@@ -97,135 +109,169 @@ export class ZeroDivisionError extends ArithmeticError {
-      Error.captureStackTrace(this, ZeroDivisionError);
+      Error.captureStackTrace(this, NameError);
     }
   }
 }
 
 export class UnboundLocalError extends NameError {
-  constructor(message?: string, name = "UnboundLocalError") {
-    super(message, name);
+  constructor(varName: string, loc?: Location) {
+    super(`local variable '${varName}' referenced before assignment`, loc);
+    this.name = "UnboundLocalError";
     if (Error.captureStackTrace) {
-      Error.captureStackTrace(this, ZeroDivisionError);
+      Error.captureStackTrace(this, UnboundLocalError);
     }
   }
 }
 
 export class RuntimeError extends Exception {
-  constructor(message?: string, name = "RuntimeError") {
-    super(message);
-    this.name = name;
+  constructor(message?: string, loc?: Location) {
+    super(message, "RuntimeError", loc);
     if (Error.captureStackTrace) {
-      Error.captureStackTrace(this, ZeroDivisionError);
+      Error.captureStackTrace(this, RuntimeError);
     }
   }
 }
 
 export class RecursionError extends RuntimeError {
-  constructor(message?: string, name = "RecursionError") {
-    super(message);
-    this.name = name;
+  constructor(loc?: Location) {
+    super("maximum recursion depth exceeded", loc);
+    this.name = "RecursionError";
     if (Error.captureStackTrace) {
-      Error.captureStackTrace(this, ZeroDivisionError);
+      Error.captureStackTrace(this, RecursionError);
     }
   }
 }
 
 export class SyntaxError extends Exception {
-  constructor(message?: string, name = "SyntaxError") {
-    super(message);
-    this.name = name;
+  constructor(message?: string, loc?: Location) {
+    super(message == undefined ? `invalid syntax` : message, "SyntaxError", loc);
     if (Error.captureStackTrace) {
-      Error.captureStackTrace(this, ZeroDivisionError);
+      Error.captureStackTrace(this, SyntaxError);
     }
   }
 }
 
 export class IndentationError extends SyntaxError {
-  constructor(message?: string, name = "IndentationError") {
-    super(message);
-    this.name = name;
+  constructor(message?: string, loc?: Location) {
+    super(message == undefined ? `unexpected indent` : message, loc);
+    this.name = "IndentationError";
     if (Error.captureStackTrace) {
-      Error.captureStackTrace(this, ZeroDivisionError);
+      Error.captureStackTrace(this, IndentationError);
     }
   }
 }
 
 export class TypeError extends Exception {
-  constructor(message?: string, name = "TypeError") {
-    super(message);
-    this.name = name;
+  constructor(message?: string, loc?: Location) {
+    super(message, "TypeError", loc);
     if (Error.captureStackTrace) {
-      Error.captureStackTrace(this, ZeroDivisionError);
+      Error.captureStackTrace(this, TypeError);
+    }
+  }
+}
+
+export class TypeMismatchError extends Exception {
+  constructor(expect: string, got: string, loc?: Location) {
+    super(`Expected type '${expect}'; got type '${got}'`, "TypeMismatchError", loc);
+    if (Error.captureStackTrace) {
```

Comment:
I think we should distinguish _static errors_ and _dynamic errors_, because a stack trace on a static error doesn't make much sense. Also, shouldn't these relate to BaseException somehow?

There's a lot of duplicated code here. What does `if (Error.captureStackTrace)` mean? I'm generally a little confused about these patterns. Maybe I missed a link to some discussion about why things are don this way?

Reply:
I also think we might want to save the values of _all_ the fields in the error. That way a fancier front-end rendering could have access to those values. and in a type mismatch, it's probably useful to keep around the whole Type rather than pre-converting it to a string before calling the constructor (since the front-end might want to distinguish that as well, especially with e.g. list types that are compound structures).

Reply:
Editing...